rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,716 @@
+
+The new `bound-keypair` method improves this situation in two primary ways: one
+fewer resource needs to be generated (the secret value), and maintenance can be
+performed simply by adjusting values in the resource. If a bot fails, its rejoin
+counter can be incremented easily.
+
+As an example, we can consider provisioning several bots. We'll need to account
+for future overrides so we can fix a single bot in the future if needed:
+
+``` hcl
+locals {
+  nodes = toset(["foo", "bar", "baz"])
+}
+
+resource "teleport_bot" "example" {
+  name = "example"
+  roles = ["access"]
+}
+
+variable "bot_rejoin_overrides" {
+  type = map(number)
+  default = {
+    foo = 5
+  }
+}
+
+resource "teleport_provision_token" "example" {
+  for_each = local.nodes
+
+  version = "v2"
+  metadata = {
+    name = "example-${each.key}"
+  }
+  spec = {
+    roles = ["Bot"]
+    bot_name = teleport_bot.example.name
+    join_method = "bound-keypair"
+
+    bound_keypair = {
+      rejoining = {
+        # look up node-specific count in the rejoin overrides map, default to 2
+        total_rejoins = lookup(var.bot_rejoin_overrides, each.key, 2)
+      }
+    }
+  }
+}
+
+resource "aws_instance" "example" {
+  for_each = teleport_provision_token.example
+
+  ami           = "ami-12345678"
+  instance_type = "t2.micro"
+
+  user_data = templatefile("${path.module}/user-data.tpl", {
+    initial_join_token = each.value.status.bound_keypair.initial_join_secret
+  })
+}
+```
+
+In this example, if node `bar` uses its 2 renewals, we can add a new entry for
+it in `bot_rejoin_overrides` and its `ProvisionToken` will be updated to allow
+additional renewals.
+
+#### Challenge Ceremony
+
+The challenge ceremony will take inspiration from several existing join methods.
+We can create an interactive challenge similar to TPM joining and present a
+challenge containing a nonce. To avoid completely implementing our own
+authentication ceremony, clients can use `go-jose` to marshal and sign a JWT
+which can then be verified easily on the server.
+
+TODO: This needs significant further elaboration and feedback.
+
+#### Client-Side Changes in `tbot`
+
+Bots should be informed of their number of remaining rejoins. There's a few
+methods by which we could inform bots of their remaining rejoins:
+
+1. (Recommended) Heartbeats: bots submit heartbeats at startup and on a regular
+   interval. It would be trivial to include a remaining rejoin counter in the
+   (currently empty) heartbeat response.
+
+2. Certificate field: we could include the number of remaining rejoins in a
+   certificate field.
+
+3. New RPC: we could add a new RPC for bots to fetch this, alongside any other
+   potentially useful information.
+
+4. We could grant bots permission to view their own join tokens. There is
+   precedent here as bots can view e.g. their own roles without explicitly
+   having RBAC permissions to do so.
+
+The remaining rejoin counter should then be exposed as a Prometheus metric to
```

Comment:
Perhaps the last time it (re)joined would be a good addition to the metrics as well.

Reply:
It's a bit tricky to publish date fields directly in prometheus, but I think we could publish a metric like this:
```
teleport_bot_bound_keypair_joins{bot_name="example"} 1
```

You could then alert against it with a query like this:
```
sum by (user) (teleport_bot_bound_keypair_joins[1h]) > 1
```

Prometheus should count each increase in the field and will account for counter resets if auth restarts. I think this would meet your needs?

Reply:
Perhaps a good middle-ground would be `teleport_bot_rejoined_seconds` where it has an ever increasing value with the number of seconds since the last rejoin.
Creating an alert would be as simple as checking if the value is less than 1h.

I think a metric like `system_uptime` would be comparable.